internal/v1/handler_blueprints.go

```diff
@@ -81,30 +75,67 @@ func (bb *BlueprintBody) RedactPasswords() {
 	}
 }
 
-func (u *User) setPassword(password *string) {
-	if password != nil {
-		if u.Password == nil {
-			u.Password = new(string)
-		}
-		*u.Password = *password
+// Merges Password or SshKey from other User struct to this User struct if it is not set
+func (u *User) MergeExisting(other User) {
+	if u.Password == nil {
+		u.Password = other.Password
+	}
+	if u.SshKey == nil {
+		u.SshKey = other.SshKey
 	}
 }
 
-func (u *User) setSshKey(sshKey *string) {
-	if sshKey != nil {
-		if u.SshKey == nil {
-			u.SshKey = new(string)
-		}
-		*u.SshKey = *sshKey
+// User must have name and non-empty password or ssh key
+func (u *User) hasValidValues() error {
+	nameRe := regexp.MustCompile(`\S+`)
+	validName := nameRe.MatchString(u.Name)
+	validPassword := u.Password != nil && len(*u.Password) > 0
+	validSshKey := u.SshKey != nil && len(*u.SshKey) > 0
+	isValid := validName && (validPassword || validSshKey)
+	if !isValid {
+		return fmt.Errorf("User ('%s') must have a name and either a password or an SSH key set.", u.Name)
 	}
+	return nil
 }
 
-func (u *User) hasPassword() bool {
-	return u.Password != nil && len(*u.Password) > 0
-}
+func (u *User) isValidForUpdate(h *Handlers, ctx echo.Context, blueprintId uuid.UUID, userID *Identity) error {
+	// If both password and ssh_key in request user we don't need to fetch user from DB
+	if !(u.Password != nil && len(*u.Password) > 0 && u.SshKey != nil && len(*u.SshKey) > 0) {
+		be, err := h.server.db.GetBlueprint(ctx.Request().Context(), blueprintId, userID.OrgID(), nil)
+		if err != nil {
+			return err
+		}
+		eb, err := BlueprintFromEntry(be)
+		if err != nil {
+			return err
+		}
```

Comment:
This database call, which is the most expensive operation for the whole request (few milliseconds on AWS EC2) is done in a loop for every single user. This is unnecessary, please call this just once outside of the loop and pass the `Users` pointer into the function.

Reply:
Now called once before iterating over users